.travis.yml

```diff
@@ -89,45 +114,70 @@ install:
       conda update -q conda
     fi
   - |
-    if [[ "$RUNMODE" != "docker" ]] ; then
+    if [[ "$RUNMODE" != "docker" ]] && [[ "$RUNMODE" != "singularity" ]]; then
       # Install conda packages
       ${TRAVIS_BUILD_DIR}/contrib/devenv/create.sh
       source ${TRAVIS_BUILD_DIR}/build/env/start
       ${TRAVIS_BUILD_DIR}/contrib/conda.sh
       ${TRAVIS_BUILD_DIR}/contrib/build-pybind11-in-conda.sh
       # Debugging information
       conda info -a
-    else
+    elif [[ "$RUNMODE" == "docker"  ]]; then
       $SCRIPT_RUN_PREFIX id
       $SCRIPT_RUN_PREFIX apt-get -qq update
       $SCRIPT_RUN_PREFIX apt-get -qqy install fakeroot debhelper locales \
         libreadline7 libssl1.0.0 libffi6 \
         liblapack3 liblapack-dev libhdf5-100 libhdf5-dev libnetcdf13 \
         libnetcdf-dev libscotch-6.0 libscotch-dev cython3 python3 \
         python3-numpy libpython3.6-dev python3-boto python3-paramiko graphviz
+    elif [[ "$RUNMODE" == "singularity"  ]]; then
+      pushd ${SHOME}
+      cp ${TRAVIS_BUILD_DIR}/sandbox/gas/tube/${SOLVCON_DRIVING} ${SHOME}
+      # trick to inject TRAVIS_BRANCH and TRAVIS_REPO_SLUG
+      echo '#!/bin/bash' > ${SHOME}/prepare-solvcon-dev.sh
+      echo "export TRAVIS_BRANCH=${TRAVIS_BRANCH}" >> ${SHOME}/prepare-solvcon-dev.sh
+      echo "export TRAVIS_REPO_SLUG=${TRAVIS_REPO_SLUG}" >> ${SHOME}/prepare-solvcon-dev.sh
+      cat ${TRAVIS_BUILD_DIR}/contrib/singularity/prepare-solvcon-dev.sh >> ${SHOME}/prepare-solvcon-dev.sh
+      cp ${TRAVIS_BUILD_DIR}/contrib/singularity/${SRECIPE} ${SHOME}
+      sudo singularity build ${SIMAGE} ${SRECIPE}
```

Comment:
I remember that I wanted to test singularity image building in travis.  However, since we aren't building the docker image here, we probably shouldn't do it for singularity, either.

I still think we should build (and test) the singularity image automatically.  But mixing the CI for SOLVCON and its containers here looks weird, after I see the code now.  And it makes CI slow.

Will there be a better way to handle both testing?

Reply:
My gut feeling told me we should leverage Singularity Hub, which is aim at building Singularity image. The image built by Singularity Hub could be pulled to somewhere so the travis CI could cache it and focus on running SOLVCON driving script. I need to evaluate if it is feasible.

Overall, I incline to the idea that we should not build the associated Singularity image here.

Reply:
If Singularity Hub hosts our image, in theory we can pull it in travis vm just like how we do it with the docker image.